### DISCUSSION TITLE
Some suggestion

### DISCUSSION THREAD

**Discussion by timreichen** [jcs224/deno_sessions#2] [Ideas]
I suggest the following changes:
# Separate Class for each framework
have a class for each framework instead of passing a property framework string.
```ts
const session = new OakSession(…)
```
instead of
```ts
const session =  new Session({ framework: "oak" })
```

# Framework instance instead of method  callings
pass a framework instance instead of calling `use`.
```ts
const session = new OakSession(app);
```
instead of
```ts
const session = new Session({ framework: "oak" });
app.use(session.use()(session));
```

# Pass options during new calls
```ts
const session = new OakSession(app, { path: "/", httpOnly: true, secure: false })
```
instead of
```ts
app.use(session.use()(session, { path: "/", httpOnly: true, secure: false }))
```

# Pass store instance instead of string
```ts
import { Session, SqliteStore } from "https://deno.land/x/session/mod.ts";
const store = new SqliteStore()
await store.connect()
const session = new OakSession(app, { store })
```
instead of
```ts
await session.init();
```
Maybe SqliteStore could even take a sqlite DB instance to have make it more flexible and if none is provided, create a new one.

# Minor things
- As a general suggestion, I would say that underscore properties/methods are not really necessary with typescript because we can use `private` instead. That would make the code a bit more readable.
- I think there is an export missing for `IStore` in interface.ts. Maybe call the interface `Store` and the file `store.ts` instead?
- simplify
```ts
if(Object.keys(this._sessionMemoryStore).includes(sessionId)) {
  return true;
} else {
  return false;
}
```
could be simplified to `return Object.keys(this._sessionMemoryStore).includes(sessionId)`.

references: https://github.com/oslabs-beta/dashport/issues/39

**Comment by jcs224** [comment]
Thanks @timreichen, I took many of your suggestions to heart and started a full rewrite of the library. It will take some time to get all of the original functionality back and will warrant a full version update. Here is the rewrite branch:

https://github.com/jcs224/deno_sessions/tree/rewrite

**Comment by timreichen** [comment]
That is awesome! I looked over with fresh eyes today and thought of some more things. What would you think of:
# Make subclasses of Store class
maybe add a `Store` class and make all other stores subclasses of it. This will prevent code duplication in the future and the session class can take `Store` as a type that will work with every Store.
# SqliteStore with DB instance parameter in constructor
Maybe SqliteStore (or any other db to come) should take a DB instance instead of path. That would give the flexibility to use any sqlite connection and database repo they chose. I think this. would make the stores nice flexible wrappers.
```ts
const database = new DB("./database.db")
const store = new SqliteStore(database)
```
instead of
```ts
const store = new SqliteStore('./database.db')
```
# add naming options
default to `session` but allow the user to change the table name. Same goes with the `state.session` property in `OakSession`. Would be nice to have that as default but configurable.

# remove lib dir?
Maybe there is no real need for a lib if the store would live top level or in the stores dir? Couldn't `Session` and `SessionData`  be combined into `Session`? Seems the one just uses the other as a property and nothing else.
 
# cleaner Store methods
I think it would make sense to have the methods be more like `Map`, `Set` or `Indexeddb` methods to give it more a js feel.
What would you think of
- rename `createSession` to `addSession`
- rename `getSessionById` to `getSession`
- rename `setSessionVariable` to `setSession` and take an object instead of a single variable key and value. The update of the object should be done outside the function, for example:
```ts
const session = store.getSession(sessionId);
const updatedSession = {
...session,
newKey: "newValue"
}
const session = store.setSession(sessionId, updatedSession);
```
# minor things
- I would try to avoid `any` type if possible. There are a lot of cases still.
- neat defaults:
```ts
constructor (store: Store=new MemoryStore()) {
  this.store = store
}
```
instead of
```ts
constructor (store?: any) {
    this.store = store || new MemoryStore
  }
```

**Comment by jcs224** [comment]
Thanks, great suggestions again. 

The only point I disagree on is making a generic Store class. What I'll probably do is create an interface to enforce certain functions being used for a storage driver. That will help ensure the driver is fully compatible with the library while giving driver creators ultimate flexibility how they implement it.

Overall, my goal is to push as much code out of the library users' code as much as possible. Definitely inspired by the Laravel / Rails ethos of convention over configuration.

Everything else, I'm on board with :smiley: fortunately, most of these improvements are in the guts of the library, not user-facing, so it shouldn't disrupt people who start using it.

**Comment by timreichen** [comment]
The think making it a subclass will be helpful in the future in case of shared method/properties between different store types (like a logger for example). How the subclass implements the driver is still up to the user. I think it would make life a bit easier in the future, but that is up to you.

**Comment by jcs224** [comment]
Thanks for your suggestion.

Also, with regards to TypeScript, I'm very new to it. I understand its benefits, but I only know the basics. Any help or contributions on how to improve that are welcome!
